### PR TITLE
Set Insertion Rework

### PR DESCRIPTION
**Pull request by philipaconrad**
This PR introduces _lazy key slice sorting_ for the Set datatype, similar to what was done in #4830 (Object Insertion Rework). After this change, `Set.Add()` will be an O(1) operation, and sorting of the Set type's key slice will be delayed until just-before-use, identically to how lazy key slice sorting is done for the Object type.

This will move the sorting overhead from construction-time for Sets over to eval-time, allowing much more efficient construction and use of enormous (500k+ item) Sets. In small-scale tests, this appears to be a performance-neutral change overall, while dramatically improving performance for the "large set" edge case.

This PR also introduces a 2x new benchmarks for Sets, to keep parity with what we introduced for Objects:
 - `BenchmarkSetCreationAndLookup` (Identical to `BenchmarkObjectCreationAndLookup`)
 - `BenchmarkSetMarshalJSON` (We already benchmark `String()` operations separately for Sets)

-----

Note: This change is independent of the PR for improving the set `union` builtin (#4980), but very slightly improves performance there as well (around 5% or smaller improvement), likely from improved efficiency in Set construction.

### CODE REVIEW

File: ast/term.go [hunk 1]
```diff
@@ -1411,6 +1420,14 @@ func (s *set) String() string {
 	return b.String()
 }
 
+func (s *set) sortedKeys() []*Term {
+	if s.numInserts > 0 {
+		sort.Sort(s.keys)
```
Comment (srenatus) [inline ast/term.go:1425]: [nit] We could have done this here, and not introduce the interface above, couldn't we?

```suggestion
		sort.Sort(termSlice(s.keys))
```

**Review by srenatus — DISMISSED** [review]
LGTM one nitpick 😄

File: ast/term.go (hunk 1 above)
Comment (philipaconrad) [reply ast/term.go:1425]: That is a solid suggestion, and I'm rolling that in shortly. :smiley:

**Review by srenatus — APPROVED** [review]
Thanks!

**Comment by philipaconrad** [comment]
Will update + squash-merge when green. :+1:
